packages/eslint-plugin/src/rules/strict-boolean-expressions.ts

```diff
@@ -275,6 +285,103 @@ export default createRule<Options, MessageId>({
       if (assertedArgument != null) {
         traverseNode(assertedArgument, true);
       }
+      if (isArrayMethodCallWithPredicate(context, services, node)) {
+        const predicate = node.arguments.at(0);
+
+        if (predicate) {
+          checkArrayMethodCallPredicate(predicate);
+        }
+      }
+    }
+
+    /**
+     * Dedicated function to check array method predicate calls. Reports predicate
+     * arguments that don't return a boolean value.
+     *
+     * Ignores the `allow*` options and requires a boolean value.
+     */
+    function checkArrayMethodCallPredicate(
+      predicateNode: TSESTree.CallExpressionArgument,
+    ): void {
+      const isFunctionExpression = ASTUtils.isFunction(predicateNode);
+
+      // custom message for accidental `async` function expressions
+      if (isFunctionExpression && predicateNode.async) {
+        return context.report({
+          node: predicateNode,
+          messageId: 'predicateCannotBeAsync',
+        });
+      }
+
+      const type = checker.getApparentType(
+        services.getTypeAtLocation(predicateNode),
+      );
```

Comment:
Nice catch! This is completely unnecessary, I'm not sure what I was thinking 😆

I've updated the code to use `services.getTypeAtLocation()`, but I also reviewed the PR as it's somewhat old, and I've learned a lot since then. I found a bug of not checking type constraints, which I've addressed (and added relevant tests), so this code looks slightly different than the suggestion above.